src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_wrapper_program.cpp

```diff
@@ -0,0 +1,1141 @@
+  // )
+  const auto write_set = ensures_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(write_set));
+
+  const auto address_of_write_set =
+    addr_of_ensures_write_set_symbol_opt->symbol_expr();
+  preamble.add(goto_programt::make_decl(address_of_write_set));
+  preamble.add(goto_programt::make_assignment(
+    address_of_write_set, address_of_exprt(write_set)));
+
+  auto function_symbol =
+    library.dfcc_fun_symbol.at(dfcc_funt::WRITE_SET_CREATE).symbol_expr();
+  code_function_callt call(function_symbol);
+  auto &arguments = call.arguments();
+
+  // write set
+  arguments.emplace_back(address_of_write_set);
+
+  // max assigns clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // max frees clause size
+  arguments.emplace_back(from_integer(0, size_type()));
+
+  // replacement mode
+  arguments.emplace_back((exprt)false_exprt());
+
+  // assume_requires_ctx
+  arguments.emplace_back((exprt)false_exprt());
+
+  // assume_requires_ctx
+  arguments.emplace_back((exprt)false_exprt());
+
+  if(contract_mode == dfcc_contract_modet::CHECK)
+  {
+    // assume_ensures_ctx
+    arguments.emplace_back((exprt)false_exprt());
+
+    // assert_ensures_ctx
+    arguments.emplace_back((exprt)true_exprt());
+  }
+  else
+  {
+    // assume_ensures_ctx
+    arguments.emplace_back((exprt)true_exprt());
+
+    // assert_ensures_ctx
+    arguments.emplace_back((exprt)false_exprt());
+  }
+
+  preamble.add(goto_programt::make_function_call(call));
+
+  // call link_allocated(pre_post, caller) if in REPLACE MODE
+  if(contract_mode == dfcc_contract_modet::REPLACE)
+  {
+    auto function_symbol =
+      library.dfcc_fun_symbol.at(dfcc_funt::LINK_ALLOCATED).symbol_expr();
+    code_function_callt call(function_symbol);
+    auto &arguments = call.arguments();
+    arguments.emplace_back(address_of_write_set);
+    arguments.emplace_back(caller_write_set_symbol.symbol_expr());
+    link_allocated_caller.add(goto_programt::make_function_call(call));
+  }
+
+  // check for absence of allocation/deallocation in contract clauses
+  // ```c
+  // DECL __check_no_alloc: bool;
+  // CALL __check_no_alloc = check_empty_alloc_dealloc(write_set);
+  // ASSERT __check_no_alloc;
+  // DEAD __check_no_alloc: bool;
+  // ```
+  auto check_var = get_fresh_aux_symbol(
+                     bool_typet(),
+                     id2string(wrapper_symbol.name),
+                     "__no_alloc_dealloc_in_ensures_clauses",
+                     source_locationt{}, // TODO add property class and location
+                     wrapper_symbol.mode,
+                     goto_model.symbol_table)
+                     .symbol_expr();
+
+  postamble.add(goto_programt::make_decl(check_var));
+
+  postamble.add(goto_programt::make_function_call(code_function_callt{
+    check_var,
+    library
+      .dfcc_fun_symbol
+        [dfcc_funt::WRITE_SET_CHECK_ALLOCATED_DEALLOCATED_IS_EMPTY]
```

Comment:
You're using `at` in other places?!

Reply:
We hardly use `at` in the code base, but I don't really mind. I'd appreciate consistency.